File: codeforces/0902B_-_Coloring_a_Tree/main.py
from sys import stdin, stdout
def main():
    def dfs(i, par):
        nonlocal adj, c
        ret = 0
        if par == -1 or c[i - 1] != c[par - 1]:
            ret += 1
        for j in adj[i]:
            if j != par:
                ret += dfs(j, i)
        return ret
    from collections import defaultdict
    n = int(stdin.readline().strip())
    adj = defaultdict(list)
    p = list(map(int, stdin.readline().strip().split(' ')))
    for i, pi in enumerate(p):
        adj[pi].append(i + 2)
        adj[i + 2].append(pi)
    c = list(map(int, stdin.readline().strip().split(' ')))
    return dfs(1, -1)

stdout.write('{}\n'.format(main()))

File: codeforces/0902B_-_Coloring_a_Tree/main.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long int ll;

const int MAXN = 1e5 + 7;

vector<int> adj[MAXN];
int c[MAXN];

int dfs(int i, int par) {
  int ret = 0;
  if (par == -1 or c[i - 1] != c[par - 1])
    ret += 1;
  for (int j : adj[i]) {
    if (j != par) {
      ret += dfs(j, i);
    }
  }
  return ret;
}

int main() {
  int n;
  scanf("%d", &n);
  for (int i = 0; i < n - 1; i++) {
    int p;
    scanf("%d", &p);
    adj[p].push_back(i + 2);
    adj[i + 2].push_back(p);
  }
  for (int i = 0; i < n; i++) {
    scanf("%d", &c[i]);
  }

  printf("%d\n", dfs(1, -1));
  return 0;
}
